functions/src/definitions/userHandlers.js

```diff
@@ -275,96 +287,194 @@ async function newImageInstanceHandler({
 }) {
   const db = admin.firestore()
   let filename
+  let messageRef
+  let hasMatch = false
+  let captionHash = caption ? hashMessage(caption) : null
+
   //get response buffer
   let buffer = await downloadWhatsappMedia(mediaId, mimeType)
   const hash = await getHash(buffer)
+
+  //upload to storage
+  const storageBucket = admin.storage().bucket()
+  filename = `images/${mediaId}.${mimeType.split("/")[1]}`
+  const file = storageBucket.file(filename)
+  const stream = file.createWriteStream()
+  await new Promise((resolve, reject) => {
+    stream.on("error", reject)
+    stream.on("finish", resolve)
+    stream.end(buffer)
+  })
+  let matchType = "none" // will be set to either "similarity" or "image" or "none"
+  //check if identical image exists among instances
   let imageMatchSnapshot = await db
-    .collection("messages")
-    .where("type", "==", "image")
+    .collectionGroup("instances")
     .where("hash", "==", hash)
```

Comment:
This seems to be using AND logic. I would assume that the image hash alone matching would be enough to consider it a match? or is it intended that image + caption pairing is always required?

Reply:
Intended, cos the caption may change the meaning of the message.